File: Flatten2Dvector.cpp
class Vector2D {
    vector<int> m_flattened_vec {};
    int m_ptr {};
    int m_sz {};
public:
    Vector2D(vector<vector<int>>& vec) {
        m_ptr = 0;
        m_sz = 0;
        for(auto& el : vec)
            m_flattened_vec.insert(m_flattened_vec.end(), el.begin(), el.end());

        m_sz = m_flattened_vec.size();
    }
    
    int next() {
        return m_flattened_vec[m_ptr++];
    }
    
    bool hasNext() {
        return (m_ptr != m_sz);
    }
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D* obj = new Vector2D(vec);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

// Alternative implementation
class Vector2D {
    vector<int> m_flattened_vec {};
    vector<int>::iterator itr = m_flattened_vec.begin();

public:
    Vector2D(vector<vector<int>>& vec) {
        for(auto& el : vec)
            m_flattened_vec.insert(m_flattened_vec.end(), el.begin(), el.end());

        itr = m_flattened_vec.begin();
    }
    
    int next() {
        return *(itr++);
    }
    
    bool hasNext() {
        return (itr != m_flattened_vec.end());
    }
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D* obj = new Vector2D(vec);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: Python/medium_difficulty/Flatten2DVector.py
class Vector2D:
    def __init__(self, vec: List[List[int]]):
        self._flattened_list = []
        self._ptr = 0

        for l in vec:
            self._flattened_list.extend(l)

        self._size = len(self._flattened_list)

    def next(self) -> int:
        next_el = self._flattened_list[self._ptr]
        self._ptr += 1
        return next_el

    def hasNext(self) -> bool:
        return self._ptr != self._size
        


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(vec)
# param_1 = obj.next()
# param_2 = obj.hasNext()